### PR TITLE
list slice/extract bugs #1053

### PR DESCRIPTION
**Pull request by anuchak**
#1053

### CODE REVIEW

File: src/include/function/list/operations/list_extract_operation.h [hunk 1]
```diff
@@ -26,10 +26,12 @@ struct ListExtract {
     template<typename T>
     static inline void operation(
         ku_list_t& list, int64_t pos, T& result, ValueVector& resultValueVector) {
-        auto uint64Pos = (uint64_t)pos;
-        if (list.size < uint64Pos) {
-            throw RuntimeException("list_extract(list, index): index=" + TypeUtils::toString(pos) +
-                                   " is out of range.");
+        if (pos == 0 || abs(pos) > list.size) {
```
Comment (andyfengHKU) [inline src/include/function/list/operations/list_extract_operation.h:29]: This seems to be incorrect. You are returning without setting any value for result which leaves it into an undefined state. The right thing should be setting resultValueVector.setNull(pos or pos - 1) to make sure result is a null value.

Add test so see if you can reproduce the undefined state case.

File: src/include/function/list/operations/list_extract_operation.h [hunk 2]
```diff
@@ -26,10 +26,12 @@ struct ListExtract {
...
+            return;
+        }
+        auto uint64Pos = (pos < 0) ? (pos + list.size + 1) : (uint64_t)pos;
+        if (uint64Pos == 0) {
```
Comment (andyfengHKU) [inline src/include/function/list/operations/list_extract_operation.h:33]: what is this case stands for, will we ever go inside this if statement?

File: src/include/function/list/operations/list_extract_operation.h [hunk 3]
```diff
@@ -26,10 +26,12 @@ struct ListExtract {
...
+            uint64Pos = list.size;
```
Comment (andyfengHKU) [inline src/include/function/list/operations/list_extract_operation.h:34]: I don't see changes applied to string list extraction. Make sure their behaviour is consistent and add test.

File: src/include/function/list/operations/list_slice_operation.h [hunk 4]
```diff
@@ -30,8 +30,12 @@ struct ListSlice {
 
     static inline void operation(ku_string_t& str, int64_t& begin, int64_t& end,
         ku_string_t& result, ValueVector& resultValueVector) {
-        int64_t startIdx = (begin == 0) ? 1 : begin;
-        int64_t endIdx = (end == 0) ? str.len : end;
+        int64_t startIdx = (begin == 0) ? 1 : ((begin < 0) ? (begin + str.len + 1) : begin);
```
Comment (andyfengHKU) [inline src/include/function/list/operations/list_slice_operation.h:33]: the logic handling negative index seems to be same for list extract and slice. So refactor a list operation utils class/struct to hold this logic.

File: src/include/function/list/operations/list_slice_operation.h [hunk 5]
```diff
@@ -30,8 +30,12 @@ struct ListSlice {
...
+        int64_t endIdx = (end == 0) ? str.len : ((end < 0) ? (end + str.len + 1) : end);
+        if (endIdx < startIdx || startIdx < 0 || endIdx < 0) {
```
Comment (andyfengHKU) [inline src/include/function/list/operations/list_slice_operation.h:35]: Similar to list extraction, you seem to only modify one of the template functions. This one is for string slice but not general list slice. Make sure you modified both and test both cases. Also, don't forget to refactor/template common logic whenever possible.

A separate comment not relates to this PR (but let's change it in this PR), use substring operation to do string slice. It's always good to reuse code.

**Review by andyfengHKU — CHANGES_REQUESTED** [review]

File: src/include/function/list/operations/list_extract_operation.h (hunk 2 above)
Comment (anuchak) [reply src/include/function/list/operations/list_extract_operation.h:33]: this case is not feasible, removed

File: src/include/function/list/operations/list_extract_operation.h (hunk 3 above)
Comment (anuchak) [reply src/include/function/list/operations/list_extract_operation.h:34]: string extract cases are handled in the array extract function, only 1 case was missing which was the -ve index out of range, I've added logic and test for it

**Comment by anuchak** [comment]
This PR is being closed, Guodong will be taking over this ticket: #1053

**Pull request closed without merging**